ember-server/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala

```diff
@@ -155,18 +151,22 @@ private[server] object ServerHelpers {
         }
         .drain
 
-    Stream
-      .eval(termSignal)
-      .flatMap(terminationSignal =>
-        sg.server[F](bindAddress, additionalSocketOptions = additionalSocketOptions)
-          .map { connect =>
-            Stream
-              .resource(connect.flatMap(upgradeSocket(_, tlsInfoOpt)))
-              .flatMap(withUpgradedSocket(_))
-          }
-          .parJoin(maxConcurrency)
-          .interruptWhen(terminationSignal)
-          .drain)
-
+    // TODO: The prefetch a really dirty hack that hurts performance a bit.
+    // We want to be able to close the server channel _before_ forcing sockets to close,
+    // so it goes without saying that we need to divorce the scopes that they ultimately live on.
+    // prefetch runs things through a queue and gives us back the consuming stream, so that
+    // when the original stream interrupts, it can finalize while still letting the current
+    // one run. I think we can write a custom combinator that achieves both
+    sg.server[F](bindAddress, additionalSocketOptions = additionalSocketOptions)
+      .interruptWhen(shutdown.signal.attempt)
+      .prefetch
```

Comment:
Separately, I'm looking into improving the performance of `parJoin` and getting rid of this `prefetch` altogether, but if we're not OK with a slight performance hit we can hold this PR